### PR TITLE
[FIX] Scrapping consumable products allowing negative quantities

### PR DESCRIPTION
**Pull request by rgarnau**
**Description of the issue/feature this PR addresses**:
It is not possible to create scraps of consumable products when there isn't enough stock. It should be possible to generate scrapping, even when there's not enough product available.

**Current behavior before PR**: 
When creating a scrap with a product of type 'Consumable' if the indicated quantity is greater than the actual stock, when pressing 'Done' you get this error:
Error message "You cannot scrap a move without having available stock for "X Product". You can correct it with an inventory adjustment."

**Desired behavior after PR is merged**:
The scrap is done correctly, even when this means that negative quants are created.

The rationale behind creating scraps for consumables is that in some cases the business wants to reflect when product has been scrapped, for quality tracking purposes. Same as it makes sense to include consumables in pickings (to reflect that the consumable is being moved).

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by blaggacao** [comment]
@jco-odoo Sounds like a hit.

**Comment by jco-odoo** [comment]
Indeed, we can improve it, but the case you keep track of scrap of consumable seems rare.  @nim-odoo

**Comment by rgarnau** [comment]
@jco-odoo - Full tests were added in the PR regarding the stock_scrap. 

cc @jbeficent

**Comment by jbeficent** [comment]
:+1: technical and functional review.

**Comment by jco-odoo** [comment]
Indeed, could have been merged already.  Will check.

**Comment by mreficent** [comment]
@jco-odoo how is this?

File: addons/stock/tests/test_stock_scrap.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            'name': 'scrap - grand test',
+            'product_id': self.product_stockable.id,
+            'product_uom_id': self.ref('product.product_uom_unit'),
+            'scrap_qty': 100,
+        })
+        scrap.onchange_product_id()
+        self.assertEquals(scrap.state, 'done')
+        self.assertEquals(self.product_stockable.qty_available, 0.0)
+        with self.assertRaises(UserError):
+            scrap.sudo(self.user_demo).unlink()
```
Comment (jco-odoo) [inline addons/stock/tests/test_stock_scrap.py]: I don't think you mean unlink() here.

**Comment by sle-odoo** [comment]
it's fine by me, maybe add a comment to not forward port the tests as they have an equivalent in https://github.com/odoo/odoo/blob/11.0/addons/stock/tests/test_move.py#L2546

File: addons/stock/tests/test_stock_scrap.py (hunk 1 above)
Comment (jco-odoo) [reply addons/stock/tests/test_stock_scrap.py]: should be do_scrap

**Comment by rgarnau** [comment]
@sle-odoo @jco-odoo ready to be merged!

File: addons/stock/tests/test_stock_scrap.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        scrap = self.scrap_obj.create({
+            'name': 'scrap - grand test',
+            'product_id': self.product_stockable.id,
+            'product_uom_id': self.ref('product.product_uom_unit'),
+            'picking_id': picking.id,
+            'scrap_qty': 80,
+        })
+        picking.action_assign()
+        picking.do_transfer()
+        self.assertEquals(scrap.state, 'done')
```
Comment (jco-odoo) [inline addons/stock/tests/test_stock_scrap.py]: Weird... Where is the do_scrap?

Comment (jco-odoo) [reply addons/stock/tests/test_stock_scrap.py]: or action_validate in fact...

**Comment by jco-odoo** [comment]
Anyone still working on this or do we just skip tests?

File: addons/stock/tests/test_stock_scrap.py (hunk 2 above)
Comment (jbeficent) [reply addons/stock/tests/test_stock_scrap.py]: do_scrap is triggered automatically when you create the scrap object, right?

**Comment by jco-odoo** [comment]
Indeed, my bad.  But then validating the picking is not necessary either.

**Comment by jbeficent** [comment]
Agreed.

**Comment by jbeficent** [comment]
Do you want us to remove the validation of the picking then?

File: addons/stock/tests/test_stock_scrap.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    def test_stock_scrap_01(self):
+        """ Standalone scrap a stockable product """
+        scrap = self.scrap_obj.create({
+            'name': 'scrap - grand test',
+            'product_id': self.product_stockable.id,
+            'product_uom_id': self.ref('product.product_uom_unit'),
+            'scrap_qty': 100,
+        })
+        scrap.onchange_product_id()
```
Comment (jco-odoo) [inline addons/stock/tests/test_stock_scrap.py]: onchange does not make sense then

**Comment by jco-odoo** [comment]
Maybe put the assert before the validation of the picking to make clear it
has nothing to do with it.

**Comment by rgarnau** [comment]
@jco-odoo Changes have been applied in the tests.

**Comment by jco-odoo** [comment]
I adapted the test as it is doing do_scrap() in the create() and updated the cla. Merged in https://github.com/odoo/odoo/commit/46129e753051816acb6bb356212bf6f95c243c17

**Pull request closed without merging**